### DISCUSSION TITLE
Set constraints for a specific file

### DISCUSSION THREAD

**Discussion by amezin** [renovatebot/renovate#29520] [Request Help] [answered]
### What would you like help with?

I would like help with my configuration

### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us which platform (GitHub, GitLab, etc) and which version of Renovate.

Bitbucket Server, Docker image ghcr.io/renovatebot/renovate:37.388.1-full

### Please tell us more about your question or problem

I want to update a reference to a pypi package using regex manager (the version is in a text file, not understood by any Python managers). And the package versions must be compatible with old Python (3.6 in my case). I can achieve that with

```json
  "constraints": {
    "python": "==3.6"
  },
  "constraintsFiltering": "strict"
```

But Renovate doesn't allow me to put `constraints` into `packageRules`. Is there a way to limit constraints to the specific file?

Example: https://github.com/amezin/renovate-regex-pypi-constraint (the pypi package itself is called "regex", may be confusing)

### Logs (if relevant)

If I put `constraints` into `packageRules`:

<details><summary>Logs</summary>

```
 INFO: Repository has invalid config (repository=local)
       "error": {
         "validationSource": "renovate.json",
         "validationError": "The renovate configuration file contains some invalid settings",
         "validationMessage": "Invalid configuration option: packageRules[0].constraints.python",
         "message": "config-validation",
         "stack": "Error: config-validation\n    at mergeRenovateConfig (/usr/local/renovate/lib/workers/repository/init/merge.ts:198:19)\n    at getRepoConfig (/usr/local/renovate/lib/workers/repository/init/config.ts:14:12)\n    at initRepo (/usr/local/renovate/lib/workers/repository/init/index.ts:55:12)\n    at Object.renovateRepository (/usr/local/renovate/lib/workers/repository/index.ts:62:14)\n    at attributes.repository (/usr/local/renovate/lib/workers/global/index.ts:218:11)\n    at start (/usr/local/renovate/lib/workers/global/index.ts:203:7)\n    at /usr/local/renovate/lib/renovate.ts:18:22"
       }
```

</details>

**Comment by rarkins** [comment]
This validation doesn't look right. Please add a minimal reproduction

**Reply by amezin** [reply]
Done, in the same repository: https://github.com/amezin/renovate-regex-pypi-constraint/commit/5ba5e00d6b9bf45a7f643bd2dbc2528a97f1d55d

Now it's failing: https://github.com/amezin/renovate-regex-pypi-constraint/issues/3

**Reply by rarkins** [reply, marked as answer]
- https://github.com/renovatebot/renovate/pull/29533
